concordium-consensus/src/Concordium/GlobalState/Persistent/ReleaseSchedule.hs

```diff
@@ -0,0 +1,415 @@
+            | otherwise = (rs{rsFirstTimestamp = minTS}, accum)
+          where
+            -- Pattern match lazily in case the map is empty.
+            ~((minTS, accs), newMap) = Map.deleteFindMin rsMap
+
+-- |A set of accounts represented by 'AccountIndex'.
+newtype AccountSet = AccountSet {theAccountSet :: Set.Set AccountIndex}
+    deriving (Serialize)
+
+instance MonadBlobStore m => BlobStorable m AccountSet
+instance Applicative m => Cacheable m AccountSet
+
+-- |A release schedule for the P5 protocol version.  This uses a 'Trie.Trie' mapping 'Timestamp's
+-- to sets of accounts.
+data NewReleaseSchedule = NewReleaseSchedule
+    { -- |The first timestamp at which a release is scheduled (or the maximum possible timestamp if
+      -- no releases are scheduled).
+      rs5FirstTimestamp :: !Timestamp,
+      -- |A map recording the first release time for each account with a pending release.
+      -- An account should occur at most once in the map.
+      rs5Map :: !(Trie.TrieN BufferedFix Timestamp AccountSet)
+    }
+
+instance MonadBlobStore m => BlobStorable m NewReleaseSchedule where
+    storeUpdate NewReleaseSchedule{..} = do
+        (pmap, newMap) <- storeUpdate rs5Map
+        let !p = do
+                put rs5FirstTimestamp
+                pmap
+        let !rs = NewReleaseSchedule{rs5Map = newMap, ..}
+        return (p, rs)
+    load = do
+        rs5FirstTimestamp <- get
+        mmap <- load
+        return $! do
+            rs5Map <- mmap
+            return $! NewReleaseSchedule{..}
+
+instance (MonadBlobStore m) => Cacheable m NewReleaseSchedule where
+    cache rs = do
+        newMap <- cache (rs5Map rs)
+        return $! rs{rs5Map = newMap}
+
+instance (MonadBlobStore m) => ReleaseScheduleOperations m NewReleaseSchedule where
+    type AccountRef NewReleaseSchedule = AccountIndex
+
+    addAccountRelease ts ai rs = do
+        (_, rs5Map) <- Trie.adjust addAcc ts (rs5Map rs)
+        return $!
+            NewReleaseSchedule
+                { rs5FirstTimestamp = min ts (rs5FirstTimestamp rs),
+                  ..
+                }
+      where
+        addAcc Nothing = return ((), Trie.Insert (AccountSet (Set.singleton ai)))
+        addAcc (Just (AccountSet accs)) = return $!! ((), Trie.Insert (AccountSet (Set.insert ai accs)))
+
+    updateAccountRelease oldts newts ai rs = do
+        (_, rsRem) <- Trie.adjust remAcc oldts (rs5Map rs)
+        (_, rs5Map) <- Trie.adjust addAcc newts rsRem
+        return $!
+            NewReleaseSchedule
+                { rs5FirstTimestamp = min newts (rs5FirstTimestamp rs),
+                  ..
+                }
+      where
+        remAcc Nothing = error "updateAccountRelease: no entry at expected release time"
+        remAcc (Just (AccountSet accs)) =
+            return $!
+                let accs' = Set.delete ai accs
+                 in if Set.null accs' then ((), Trie.Remove) else ((), Trie.Insert (AccountSet accs'))
```

Comment:
That's a funny single space indent.

Reply:
I don't like it, but the formatter does it.